### ISSUE TITLE
Resolve ambiguity of tuple return values

### ISSUE THREAD

**Issue by ghost** [dotnet/csharplang#2202]
I tried this:
```
(int, string) Foo() => (1, "test");
        
(int, string, bool) Foo() => (1, "test", false);
```

C# considers these overloads ambiguous, although they can be not in this example:
```
   var (x, y) = Foo();
   var (x, y, z) = Foo(); 
```

Tuples change the way that functions return values, and the way we read them (deconstruction), so I think tuples deserve some new overload resolution rules to make the most of them.

**Comment by jmarolf** [comment]
Return values are never considered when calculating overload resolution.  The code below will also fail to compile due to ambiguous methods.

```C#
int Foo() => 1;

string Foo() => "test";

void M()
{
    int x = Foo();
    string y = Foo();
}
```

**Comment by MohammadHamdyGhanem** [comment]
@jmarolf
    That is true becuase you can use type inference:
```
    var x = Foo();
    var y = Foo();
```

and in some cases c# can do implicit type conversion. 
This can't happen wthi tupples with different lengthes. erturned tuples are in fact a replacement for out params list which are not ambiguous, so, I think c# should tret returned tuples that way, to prevent us from using out params in such ambiguous situations!

**Comment by ufcpp** [comment]
https://github.com/dotnet/csharplang/issues/1587

**Comment by Scott-Caldwell** [comment]
> @jmarolf
> That is true becuase you can use type inference:

For what it's worth, you can do the same thing with tuple returns as well. These are perfectly valid:

```csharp
(int, string) Foo() => (1, "test");

var x = Foo();
(var y, var z) = Foo();
```

Taking tuple returns as a special case here doesn't make much sense to me. We'd run into the same problems here as a broader proposal to allow overloads based on return type in general.

**Comment by MohammadHamdyGhanem** [comment]
@Scott-Caldwell
Using discards or optional params can caus ambiguity somtimes, So, C# gives an error about such cases. That can be done in your case, so C# should give an ambiguity error and tells you to use the deconstruction syntax to resolve it.
Tuple returns are not just a case, but a new addition to the language, that makes you write a type list as a return value. It is illogical to prevent us from wtriting two overloads while they have different return (param list)! I see it as an unnecessary limitation, that forces us to revert back to out params in such situations.

**Comment by jmarolf** [comment]
@MohammadHamdyGhanem 	you are welcome to your opinion.  The point is that implementing this proposal and #1587 are the same amount of work.  In order to disambiguate cases in return types for tuples so that the compilation error messages are nice we will need to understand overload resolution on return types in general.

**Comment by MohammadHamdyGhanem** [comment]
@jmarolf 
I'm not against to return value-based overloading, but it seems too complicated regarding deconstructed return value syntax, which resolves itself clearly in most cases. So, it it logical to start with the obvious case, then ask for more.

**Comment by HaloFour** [comment]
@MohammadHamdyGhanem 

The overload rules predate type inference.  Overloading by return value is complicated and unnecessary.  You have to consider all of the cases where the method is invoked without assigning the return value to anything, or worse, called in an expression used within another expression where there are multiple potential overloads.  It's not worth it, and allowing it would have a ripple effect across languages in the .NET ecosystem which certainly wouldn't be able to handle it either.

**Comment by MohammadHamdyGhanem** [comment]
@HaloFour
Fair enough. Although maybe there is a workaround. C# can give ambiguity error in such complex situation, and warn the programmer that using return value overloading is limited to explicit assignments only, otherwise it will cause ambiguity. Maybe it can happen someday in the future!
Thanks.

**Comment by theunrepentantgeek** [comment]
That would permit the creation of methods that could _only_ be used for assignment, and not used anywhere else.

Given this code
```
Foo foo = FindFoo(42);
Write(foo);
```

It would be _extremely odd_ if the compiler complained about this refactoring:
```
Write(FindFoo(42));
```

**Comment by yaakov-h** [comment]
There are many other fun cases that would come into effect, such implicit conversions, implicit casts, and integer extensions.

For example:

```csharp
(short, byte) Foo() => throw null;
(byte, short) Foo() => throw null;

(int x, int y) = Foo(); // Which one did we call?
```

**Comment by MohammadHamdyGhanem** [comment]
@theunrepentantgeek
I disagree with your example, Passing the return value to a param is some sort of assignment, and can be resolved in most cases. Besides dealing with tuples is easier, becuase there is already limitations on passing tuples to a generic Tuple param, so the Tuple param will have a specific signature.

**Comment by MohammadHamdyGhanem** [comment]
@yaakov-h
Let me rephrase my proposal:
consider the returned Tuple definition as if it is a param list definition and apply the same exact overload resolution rules. So, if you want your answer, convert tour example to:
```
  void Foo(short x, byte y) => through null;
  void Foo(byte x, short y) => through null;
```

Now if you try:
`Foo(1, 2);`
C# will give you this error:
> The call is ambiguous between the following methods or properties: 'Form1.Foo(short, byte)' and 'Form1.Foo(byte, short)'

This is exactly what should happen with the returned tuple overloads.

**Comment by MohammadHamdyGhanem** [comment]
By the way, the reason that made me think of this, is the odd situation where you have to use out params to declare a class deconstructor to use it in a tuple-like deconstruction syntax! You add tuple to make the decrease to out params less, but in contrary, it made it increase! This is because if we wrote a deconstructor that returns a tuple we will not be apple to overload it!

**Comment by YairHalberstadt** [comment]
Closing as duplicate of #1587
